api/src/sockets/kots/DeploySocket.ts

```diff
@@ -114,99 +115,137 @@ export class KotsDeploySocketService {
-              }
-            }
-          } catch (err) {
-            console.log("Velero restore failed");
-            console.log(err);
-          }
-          break;
-
-        case "failed":
-          logger.warn(`Restore ${app.restoreInProgressName} falied`);
-          // TODO
-          break;
-
-        default:
-          const deployedAppVersion = await this.kotsAppStore.getCurrentVersion(app.id, clusterSocketHistory.clusterId);
-          const maybeDeployedAppSequence = deployedAppVersion && deployedAppVersion.sequence;
-          if (maybeDeployedAppSequence! > -1) {
-            const deployedAppSequence = Number(maybeDeployedAppSequence);
-            const cluster = await this.clusterStore.getCluster(clusterSocketHistory.clusterId);
-            try {
-              const desiredNamespace = ".";
-              const rendered = await app.render(app.currentSequence!.toString(), `overlays/downstreams/${cluster.title}`);
-              const b = new Buffer(rendered);
+        const cluster = await this.clusterStore.getCluster(clusterSocketHistory.clusterId);
+        try {
+          await this.handleRestoreInProgress(app, cluster);
+        } catch (err) {
+          logger.warn("Failed to handle restore in progress");
+          logger.warn(err);
+        }
+      }
+    }
+  }
 
-              const kotsAppSpec = await app.getKotsAppSpec(cluster.id, this.kotsAppStore);
+  async handleRestoreInProgress(app: KotsApp, cluster: Cluster): Promise<void> {
+    if (!app.restoreInProgressName) {
+      return;
+    }
+  
+    switch (app.restoreUndeployStatus) {
+    case UndeployStatus.InProcess:
+      // undeploy in process, continue loop
+      break;
+  
+    case UndeployStatus.Completed:
+      await this.handleUndeployCompleted(app, cluster);
+      break;
+  
+    case UndeployStatus.Failed:
+      logger.warn(`Restore ${app.restoreInProgressName} falied`);
+      // TODO
+      break;
+  
+    default:
+      // start undeploy
+      await this.undeployApp(app, cluster);
+    }
+  }
+  
+  async undeployApp(app: KotsApp, cluster: Cluster): Promise<void> {
+    logger.info(`Starting restore, undeploying app ${app.name}`);
+
+    const desiredNamespace = ".";
+    const rendered = await app.render(app.currentSequence!.toString(), `overlays/downstreams/${cluster.title}`);
+    const b = new Buffer(rendered);
+
+    const kotsAppSpec = await app.getKotsAppSpec(cluster.id, this.kotsAppStore);
+
+    // make operator prune everything
+    const args = {
+      app_id: app.id,
+      kubectl_version: kotsAppSpec ? kotsAppSpec.kubectlVersion : "",
+      namespace: desiredNamespace,
+      manifests: "",
+      previous_manifests: b.toString("base64"),
+      result_callback: "/api/v1/undeploy/result",
+      wait: true,
+    };
+
+    this.io.in(cluster.id).emit("deploy", args);
+
+    await this.kotsAppStore.updateAppRestoreUndeployStatus(app.id, UndeployStatus.InProcess);
```

Comment:
Any ideas for what to do if the operator is down when the emit above fires? In that case the db will have in_process and restarting operator and api pods won't help.

Reply:
discussed irl. gonna retry this every minute until state increases